components/webgpu/swapchain.rs

```diff
@@ -343,92 +480,107 @@ impl crate::WGPU {
+                let buffer_state = context_data.get_buffer_state(buffer_id);
+                match buffer_state {
+                    PresentationBufferState::Unassigned => {
+                        // throw away all work, because we are from old swapchain
+                        return;
+                    },
+                    PresentationBufferState::Mapping => {},
+                    _ => panic!("Unexpected presentation buffer state"),
+                }
+                if config_changed {
+                    /*
+                    This means that while mapasync was running, context got recreated
+                    so we need to throw all out work away.
+
+                    It is also possible that we got recreated with same config,
+                    so canvas should be cleared, but we handle such case in gpucanvascontext
+                    with drawing_buffer.cleared
+
+                    One more case is that we already have newer map async done,
+                    so we can replace new image with old image but that should happen very rarely
+
+                    One possible solution to all problems is blocking device timeline
+                    (wgpu thread or introduce new timeline/thread for presentation)
+                    something like this is also mentioned in spec:
+
+                    2. Ensure that all submitted work items (e.g. queue submissions) have completed writing to the image
+                    https://gpuweb.github.io/gpuweb/#abstract-opdef-get-a-copy-of-the-image-contents-of-a-context
+                    */
+                    let _ = global.buffer_unmap(buffer_id);
+                    *buffer_state = PresentationBufferState::Available;
+                    return;
+                }
                 *buffer_state = PresentationBufferState::Mapped;
                 let presentation_buffer =
-                    GPUPresentationBuffer::new(global, buffer_id, buffer_size);
-                let old_presentation_buffer = present_data.data.replace(presentation_buffer);
+                    GPUPresentationBuffer::new(global, buffer_id, image_desc.buffer_size());
+                let Some(swap_chain) = context_data.swap_chain.as_mut() else {
+                    return;
+                };
+                let old_presentation_buffer = swap_chain.data.replace(presentation_buffer);
                 let mut txn = Transaction::new();
                 txn.update_image(
-                    present_data.image_key,
-                    present_data.image_desc,
-                    present_data.image_data.clone(),
+                    context_data.image_key,
+                    context_data.image_desc.0,
+                    context_data.image_data.clone(),
                     &DirtyRect::All,
                 );
                 webrender_api
                     .lock()
                     .unwrap()
                     .send_transaction(webrender_document, txn);
```

Comment:
One last thing, for the HTML loop integration to work, you would need I think to send a message to script for it to actually note the rendering opp, using https://github.com/servo/servo/blob/c519a2cdb68683e2c9b05e54cfd10e022228fb2c/components/script/script_thread.rs#L1752 (unless this happens already via the compositor in response to sending the webrender api call). Otherwise nothing will happen.

Reply:
> Currently the call to update_rendering_of_webgpu_canvas happens within a reflow, do you know what message triggers that reflow?

Reflow is caused by `mark_as_dirty`.

Reply:
I just looked at it, so `mark_as_dirty` adds the context to `dirty_webgpu_contexts`, but I think the reflow is triggered by something else, so there could be some liveness issue. But I suggest inside `mark_as_dirty`, you do `ScriptThread::note_rendering_opportunity`, and then you do `flush_dirty_webgpu_canvases` at https://github.com/servo/servo/blob/c519a2cdb68683e2c9b05e54cfd10e022228fb2c/components/script/script_thread.rs#L1745. Could be a follow-up as well.

Reply:
this should be part of html event loop integration followup.